### PR TITLE
fix: stablecoin allowlist + token-risk smart screening, IL protection, sqlite-vec memory repair, veto robustness

### PR DESCRIPTION
**Pull request by irfndi**
Fixes the three remaining real bugs reported against the deployed box, plus the requested direction change: the stablecoin allowlist is now optional and the engine detects risky vs. safe freeze-authority tokens via computed signals instead of manual curation.

## Bug 1 — Freeze-authority screening rejected USDC/USDT/PYUSD (corrected body)

The freeze gate rejected every freeze-authority token — including USDC, blocking SOL/USDC watch pools.

- `STABLECOIN_MINTS` is now the trusted-mint allowlist: default `USDC,USDT,PYUSD` (the three mints verified on the deploy box); `STABLECOIN_MINTS=""` disables the allowlist entirely; entries are pubkey-validated fail-closed like `WATCHLIST_POOLS`.
- Per-leg exemption: a freeze-enabled leg passes when its mint is trusted; the pool is rejected only when an untrusted leg has freeze authority. Data API `is_blacklisted` and `blacklist.checkPool` remain fail-closed and are **never** exempted.
- `FREEZE_SMART_SCREENING=true` (default `false`): untrusted freeze-enabled pools pass to the quality pipeline (pre-filters / confidence / risk gates) with a recorded warning instead of a hard reject — with an actionable rejection hint when off.
- **Smart detection without manual curation** (`JUPITER_TOKEN_RISK_ENABLED`, default `true`): untrusted freeze legs consult the Meteora Data API `is_verified` field and Jupiter's Tokens API V2 (`audit.isSus` — Jupiter's aggregated RugCheck+Blockaid flag — and `isVerified` + Organic Score): verified → exempt with warning, `isSus` → hard reject, unknown → smart-screening/strict path. Advisory and fail-open (fetch failures serve cached/stale signals, never block the cycle). Also consumes previously-unused Data API `token_x/y.is_verified`.
- Evidence: `bench/decision-loop-screening.test.ts` (16 tests incl. USDC-leg exemption, empty-allowlist rejection, smart pass-through, datapi-verified/jupiter-verified/isSus paths), `bench/token-risk-service.test.ts` (7 tests: parse semantics, fail-open, TTL cache, keyless header handling, 100-mint chunking). Live-verified against `api.jup.ag/tokens/v2` (keyless) and `api.rugcheck.xyz`: USDC/USDT/JupUSD/cbBTC/LBTC/MU all carry freeze authority yet are verified — exactly the case this fixes.

## Bug 2 — Native bindings never loaded; memory silently dead (corrected body)

Root cause was two-layered, deeper than reported:

- **Primary**: `vec_memory` declared `+pnlUsd REAL, +confidence REAL` — sqlite-vec 0.1.9 auxiliary columns accept only TEXT/INTEGER/DOUBLE/BLOB, so `CREATE VIRTUAL TABLE` failed on **every platform, every install** (misleading `chunk_size must be a non-zero positive integer`; `IF NOT EXISTS` swallows it when the module is absent). Memory recall/recording has been dead since inception; positive tests skip-returned so CI never caught it. Fixed to `DOUBLE` + regression test.
- **Secondary**: on macOS `Database.setCustomSQLite` only probed Homebrew — machines without it couldn't load extensions at all (Bun's bundled SQLite has no extension support; verified Apple's `/usr/lib` sqlite is not extension-capable, so doctor now fails loudly pointing at `brew install sqlite`). Added MacPorts candidate.
- `prism doctor` now has a `memory` check that **fails (exit 1)** with platform-specific remediation when vector memory can't load, and a WARN-only `native-bindings` check explaining the `bigint: Failed to load bindings` line is the harmless `bigint-buffer` pure-JS fallback (identical results, no action).
- Bundle hardening: `sqlite-vec` JS is now inlined into `dist/cli` (`noExternal`) so the npm load path fails gracefully into the `PRISM_VEC0_PATH`/embedded chain on clean machines instead of crashing on a bare import.
- Evidence: `bench/doctor-vec.test.ts` (8 tests — failure shape deterministic via mocked bundle surface), `.omo/manual-qa/t4-doctor-bundle-smoke.md` (7 darwin-arm64 bundle configurations: bug reproduction with FAIL + exit 1 on the pre-fix schema, green after the fix, binary-level aux-column bisection table).

## Bug 3 — OpenClaw gateway veto fetch failed every cycle (corrected body)

- The real gateway error now surfaces: `UnknownException` cause is unwrapped (`underlyingErrorMessage`, shared in `engine/errors.ts`; adapter-service deduped to it) so logs show e.g. `Gateway 1008: ...` / `Gateway chat run timeout` instead of the wrapper string.
- The per-cycle spam is throttled to one warn per pool per `AGENT_PROPOSAL_STALE_MS` window (suppressed occurrences drop to debug).
- The prime root cause is eliminated: the gateway transport is no longer selected when `AGENT_GATEWAY_TOKEN` is empty (detection only probes the pre-auth WebSocket upgrade, so it previously selected a scope-less transport that failed every per-decision re-handshake). Explicit `AGENT_RUNTIME=openclaw` + empty token → one actionable startup warning and veto review disabled; `auto` falls through silently.
- `getStatus().connected` is now truthful (reported `true` regardless of failed startup connect before). Fail-open semantics unchanged — veto absence never alters decisions. HTTP pull interfaces untouched.
- Evidence: `bench/gateway-transport.test.ts` (+3: ack-error surfaces `Gateway ${code}: ${message}`, chat run timeout, protocol<4 handshake), `bench/agent-service.test.ts` (+10: selectTransport token gating, connectReviewTransport truthfulness).

## Strategic direction — smart detection without manual whitelist

The requested approach (engine smart enough to detect safe tokens and protect against IL) ships in layers; this PR covers everything safe to ship now:

- **Allowlist optional** — `STABLECOIN_MINTS=""` disables it; the token-risk overlay + smart screening adjudicate everything else by computed signals.
- **Fee-vs-IL computed before every position** — new `[fee-il-gate]` ENTER floor: enters only when the pool's fee/IL ratio ≥ `MIN_FEE_IL_RATIO` (1.2); the ratio is never unknown (0–20), so it fails closed on 0.
- **Anomaly protection on held positions** — IL-dominance fast EXIT: when a position is out of range and its impermanent loss (vs the real on-chain HODL benchmark) exceeds cumulative claimed fees × `IL_DOMINANCE_EXIT_FACTOR` (2) and `IL_DOMINANCE_MIN_USD` (5), it exits immediately via the wave-15 fast-exit seam (risk-override, position-scoped) and fires an `il_dominance` Telegram alert. Pre-v16 positions without entry data fail open.
- ENTER also blocks on Jupiter hard-risk signals (`audit.isSus`, `organicScoreLabel: "low"`) — `[token-risk]` audited.

**Deferred (deliberate):** auto-converting position assets to stable/SOL on an hourly/daily cadence needs an exit+swap treasury loop and a rotation strategy — the foundations exist on main (Jupiter swap seam, IL-vs-HODL benchmark, fast-exit pipe) and this is the next iteration; shipping autonomous asset rotation inside a hotfix PR would be reckless in live mode. Also deferred: direct Rugcheck/Blockaid APIs as v2 corroboration (Blockaid is keyed-only; Jupiter's aggregated audit covers the same surface in one batched request); the `jup` CLI (`@jup-ag/cli`) and `npx skills add jup-ag/agent-skills --skill integrating-jupiter` are agent-harness tooling (install them in your harness if useful — they are not engine dependencies).

## Deploy notes

- USDC/USDT/PYUSD exemption is automatic (default allowlist). To disable: `STABLECOIN_MINTS=""`.
- Optional: `FREEZE_SMART_SCREENING=true`; IL protection is on by default (`IL_PROTECTION_ENABLED=true`, `IL_DOMINANCE_EXIT_FACTOR=2`, `IL_DOMINANCE_MIN_USD=5`); token-risk overlay on by default (`JUPITER_TOKEN_RISK_ENABLED=true`, 30-min signal cache; set `JUPITER_API_KEY` for higher rate limits — keyless works).
- `prism doctor` now exits 1 when vector memory is dead — install an extension-capable SQLite (`brew install sqlite` / `apt-get install libsqlite3-0`) to restore memory.
- Cloudflare: this PR registers the `il_dominance` alert type (API allowlist + Telegram label) — merge triggers the deploy-cloudflare workflow.

## Verification

- Full suite: **1068/1068 tests, 92 files** (+93 vs main baseline) — `bun run test`
- Coverage: 81.28% statements / 75.82% branches / 77.76% functions / 81.89% lines (thresholds 75/60/75/75) — enforced gate
- `bun run lint` (tsc --noEmit + oxlint): exit 0 · `bun run format:check`: clean · `bun run build`: green · `cd cloudflare && bun run typecheck`: exit 0
- Hands-on: `prism doctor` loud-fail + green-path smoke on source and on a darwin-arm64 bundle (7 configurations, reproduction → fix transcripts in the smoke doc)

## Summary by CodeRabbit

* **New Features**
  * Added a stablecoin allowlist–driven, configurable freeze-authority screening with an optional pass-through mode.
  * Introduced Jupiter token-risk overlay gating for entry decisions (advisory, fail-open when signals are unavailable).
  * Added IL protection with fee-to-IL entry gating and IL-dominance fast-exit, including `il_dominance` alerting.
  * Expanded `doctor` checks for vector-memory availability and native bindings status.
* **Bug Fixes**
  * Improved review-transport selection/connection reporting and gateway error surfacing.
* **Documentation**
  * Updated configuration examples and guides for freeze screening, IL protection, token-risk overlay, and the revised risk-gate behavior.

### CODE REVIEW

File: engine/token-risk-service.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    const apiKey = process.env.JUPITER_API_KEY?.trim();
+    if (apiKey) request.apiKey = apiKey;
+    if (options.fetchImpl !== undefined) request.fetchImpl = options.fetchImpl;
+    try {
+      const fetched = await fetchTokenRisks(missing, request);
+      const fetchedAt = Date.now();
+      for (const [mint, signal] of fetched) {
+        cache.set(mint, { signal, fetchedAt });
+        result.set(mint, signal);
+      }
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/token-risk-service.ts]: Fixed in 3888626. `consultTokenRisks` now negatively caches mints omitted by a successful refresh (fresh timestamp stops the per-cycle re-queries) and removes stale served entries from the result — a revoked verification no longer keeps exempting the pool. Stale serving happens only on fetch failure. Locked by new tests (8)–(9) in `bench/token-risk-service.test.ts` (revocation + never-returned negative cache).

[the remark this answers is not in the document]
Comment (irfndi) [reply engine/token-risk-service.ts]: Addressed in 3888626 — the same rework as the adjacent thread: requested-but-absent mints on a successful refresh get a negative-cache entry (stops the repeated requests) and are dropped from the served result, so a revoked or absent `isVerified` can no longer bypass screening; stale values are served only on the error path. New revocation + negative-cache tests: `bench/token-risk-service.test.ts` (8)–(9).

File: engine/program.ts [hunk 2]
```diff
@@ -2943,10 +3189,25 @@ export const program = Effect.gen(function* () {
...
-                  });
+                  const message = underlyingErrorMessage(err);
+                  // Throttle the warn to the veto-warning cadence (vetoWarningThrottle,
+                  // per pool, agentProposalStaleMs) so a persistently failing runtime
+                  // logs once per window instead of per decision per cycle; suppressed
+                  // occurrences stay visible at debug. Fail-open is preserved: this
+                  // always resolves to null and never alters the decision.
+                  const lastWarn = vetoWarningThrottle.get(poolAddress) ?? 0;
+                  if (now - lastWarn > config.agentProposalStaleMs) {
+                    vetoWarningThrottle.set(poolAddress, now);
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts]: Fixed in 3888626: throttle eligibility is computed once (`vetoWarnEligible` in the catchAll, which owns the single read+set). The memory-warning block reuses it (`else if (vetoFetchFailed && vetoWarnEligible)`) instead of re-reading the map, so the warn log and the memory warning now fire together exactly once per `AGENT_PROPOSAL_STALE_MS` window.

File: engine/agent-service.ts [hunk 3]
```diff
@@ -266,11 +267,25 @@ function selectTransport(
...
-      token: config.agentGatewayToken,
-      timeoutMs: config.agentPromptTimeoutMs,
-    });
+    // The gateway transport authenticates with the shared token; with no token the
+    // connection loses its operator scopes and every per-decision re-handshake fails
+    // (detection only probes the pre-auth WS upgrade, so it cannot catch this). Skip
+    // selection here. Warn once — selectTransport runs only at startup — and only when
+    // the user explicitly chose the openclaw runtime; `auto` falls through silently so
+    // it does not spam when other transports (or none) apply.
+    if (config.agentGatewayToken.trim() !== "") {
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/agent-service.ts]: Hardened in 3888626 as `(config.agentGatewayToken ?? "").trim() !== ""`. The suggested `config.agentGatewayToken?.trim() !== ""` was deliberately not applied verbatim: with `undefined` it yields `undefined !== "" → true`, which would construct the GatewayTransport with no token — the exact failure mode this guard prevents. The field is a required string on `AppConfig` (engine/config-service.ts), so this is defense-in-depth, not a live-crash fix.

File: engine/program.ts [hunk 4]
```diff
@@ -2265,6 +2378,43 @@ export const program = Effect.gen(function* () {
...
+          // price; the HODL benchmark is the real on-chain entry legs.
+          const hodlValueUsd = computeHodlValueUsd(
+            pos.entryAmountXUsd,
+            pos.entryAmountYUsd,
+            pos.entryPriceUsd,
+            pool.currentPrice,
+          );
+          if (hodlValueUsd !== null) {
+            const ilUsd = hodlValueUsd - pos.currentValueUsd;
+            const feesClaimedUsd = pos.cumulativeFeesClaimedUsd;
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts:2419]: Investigated — no change needed. The `ilDominance` local is only used to CONSTRUCT the decision (its pre-check and the `else if (ilDominance)` branch); no downstream categorization post-processes the decision against raw metrics — the EXIT cooldown path (`computeCooldownForExit`) already inspects the FINAL reasoning string, and the IL-dominance reasoning deliberately matches none of its trigger substrings, so overlay-modified decisions are categorized consistently.

File: .env.example [hunk 5]
```diff
@@ -127,12 +127,35 @@ DISCOVERY_MIN_FEE_RATIO=1.5
 
 # Dump full pool state snapshots to DB every cycle (paper only)
 ENABLE_SNAPSHOT_CAPTURE=false
+# Stablecoin mints trusted by depeg detection + freeze screening (allowlist).
+# Unset = USDC, USDT, PYUSD (default). Set empty (STABLECOIN_MINTS=) to disable
+# the allowlist entirely; every mint must be a valid Solana pubkey (fail-closed).
 STABLECOIN_MINTS=
```
[the remark this answers is not in the document]
Comment (irfndi) [reply .env.example]: Fixed in eb5c7fd: `.env.example` now carries the default allowlist value (matching the `loadConfig` default), so copying the example no longer disables the allowlist; the comment documents the empty-string disable switch. The setup wizards were audited and do not write this var, so the default applies.

File: engine/db.ts [hunk 6]
```diff
@@ -7,18 +7,54 @@ import { getEmbeddedVec0Path } from "./sqlite-vec-embedded.js";
...
+    // as the existence check, which matters because Apple's /usr/lib library has
+    // no on-disk file (dyld shared cache) yet still dlopens — only to fail vec0
+    // loading, which prism doctor then reports loudly as a last-resort fallback.
+    const brewPrefix = process.arch === "arm64" ? "/opt/homebrew" : "/usr/local";
+    const otherBrewPrefix = process.arch === "arm64" ? "/usr/local" : "/opt/homebrew";
+    const candidates = [
+      `${brewPrefix}/opt/sqlite/lib/libsqlite3.dylib`,
+      `${otherBrewPrefix}/opt/sqlite/lib/libsqlite3.dylib`,
+      "/usr/lib/libsqlite3.dylib",
+      "/opt/local/lib/libsqlite3.dylib",
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/db.ts]: Fixed in eb5c7fd: MacPorts `/opt/local/lib/libsqlite3.dylib` is now probed before Apple's `/usr/lib/libsqlite3.dylib` — the Homebrew prefixes stay first, and Apple's shared-cache library (dlopens but cannot load extensions) is the last-resort loud-fail path.

File: engine/agent-service.ts [hunk 7]
```diff
@@ -266,11 +267,25 @@ function selectTransport(
...
+    // selection here. Warn once — selectTransport runs only at startup — and only when
+    // the user explicitly chose the openclaw runtime; `auto` falls through silently so
+    // it does not spam when other transports (or none) apply.
+    if ((config.agentGatewayToken ?? "").trim() !== "") {
+      return new GatewayTransport({
+        url: config.agentGatewayUrl,
+        token: config.agentGatewayToken,
+        timeoutMs: config.agentPromptTimeoutMs,
+      });
+    }
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/agent-service.ts:283]: Fixed in eb5c7fd: in `auto` mode a tokenless-but-reachable gateway now falls back to the Hermes/ACP transport when available (info-logged); explicit `AGENT_RUNTIME=openclaw` keeps the warn + review-disabled semantics. Test coverage added.

File: engine/program.ts [hunk 8]
```diff
@@ -2943,10 +3189,25 @@ export const program = Effect.gen(function* () {
...
+                    logger.warn("Agent veto fetch failed", {
+                      pool: poolAddress,
+                      error: message,
+                    });
+                  } else {
+                    logger.debug("Agent veto fetch failed (throttled)", {
+                      pool: poolAddress,
+                      error: message,
+                    });
+                  }
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts:3238]: Resolved in 3888626 — veto throttle restructured (same fix as the adjacent codex thread): the catchAll computes eligibility once (`vetoWarnEligible`, single read+set) and the memory-warning block reuses it, so the warn log and memory warning fire together exactly once per `AGENT_PROPOSAL_STALE_MS` window.

File: engine/token-risk-service.ts [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+    const url = `${JUPITER_TOKENS_SEARCH_BASE_URL}?query=${query}`;
+    const res = await fetchImpl(url, {
+      headers,
+      signal: AbortSignal.timeout(timeoutMs),
+    });
+    if (!res.ok) {
+      throw new Error(`Jupiter tokens API HTTP ${res.status}`);
+    }
+    const body: unknown = await res.json();
+    if (!Array.isArray(body)) continue;
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/token-risk-service.ts]: Fixed in 10ce741: a 200 response with a non-array body now THROWS in `fetchTokenRisks` (was `continue`), so such responses route into the documented fail-open path — cached `isSus` hard-reject enforcement is preserved and malformed payloads are never negative-cached. Tests added: `fetchTokenRisks` rejects on an object body; consult serves the stale `isSus: true` signal and still refreshes normally afterward (bench/token-risk-service.test.ts (10)–(11)).

File: engine/program.ts [hunk 10]
```diff
@@ -2785,6 +2961,82 @@ export const program = Effect.gen(function* () {
...
+          // [token-risk] ENTER gate (Wave 18): Jupiter advisory overlay. Blocks
+          // entry when either leg carries a hard-risk signal — Jupiter isSus
+          // (aggregated RugCheck+Blockaid) or a "low" organic score. Advisory and
+          // fail-open: unknown, disabled or failed signals never block entry.
+          // Reuses the per-cycle token-risk cache, so a second consult costs no
+          // network call when the screening seam already fetched these mints.
+          if (!enterGateRejected && config.jupiterTokenRiskEnabled !== false) {
+            const enterRisk = yield* Effect.promise(() =>
+              consultTokenRisks([pool.tokenX, pool.tokenY], config),
+            );
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts]: Fixed in 10ce741: the `[token-risk]` ENTER gate now runs AFTER all local eligibility predicates (fee/IL floor, volume authenticity, bin utilization, TVL, weighted entry score) — the Jupiter consult fires only when a candidate ENTER survived the local gates, so locally-ineligible pools contribute zero network cost per cycle (previously up to a 10s timeout per pool when Jupiter is down). Audit-reason precedence follows: local rejections log their local reason. Locked by a zero-fetch regression test (bench/decision-loop-screening.test.ts (13)).

File: engine/db-service.ts [hunk 11]
```diff
@@ -436,29 +446,42 @@ export const DbLive = (dbPath?: string) =>
...
-                   ORDER BY distance`;
-                  const params = poolAddress
-                    ? [JSON.stringify(embedding), topK * 2, now, poolAddress]
-                    : [JSON.stringify(embedding), topK * 2, now];
-                  const rows = yield* Effect.try({
-                    try: () => queryAll<Record<string, unknown>>(db, sql, ...params),
+                   WHERE embedding MATCH ? AND k = ?
+                   ORDER BY distance`,
+                        JSON.stringify(embedding),
+                        topK * 2,
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/db-service.ts]: Addressed in 10ce741: `queryMemory` now uses an expanding KNN window — when the post-fetch filter yields fewer than `topK` in-scope rows, `k` doubles (capped at `max(topK*8, 64)`, at most three re-queries) before ranking/slicing. This restores the prior SQL pre-filter behavior in realistic cases; the genuine corner case (a table dominated by expired/other-pool rows beyond the maxK window) remains documented and fails soft. Regression test: 3 target-pool rows + 15 other-pool rows → scoped query returns exactly topK in-scope rows regardless of distance order (bench/db-memory.test.ts).

File: engine/db-service.ts [hunk 12]
```diff
@@ -436,29 +446,54 @@ export const DbLive = (dbPath?: string) =>
...
+                  // over at most three re-queries — until topK in-scope rows are
+                  // found or the cap is hit. This restores parity with the old SQL
+                  // WHERE pre-filter in realistic cases; only genuine corner cases
+                  // (a table dominated by expired/other-pool rows beyond the maxK
+                  // nearest neighbours) fail-soft with fewer rows. Recency-blend
+                  // ranking and ORDER BY distance are preserved exactly.
+                  const maxK = Math.max(topK * 8, 64);
+                  let k = topK * 2;
+                  let candidates: Record<string, unknown>[] = [];
+                  for (let expansion = 0; expansion < 4; expansion += 1) {
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/db-service.ts]: Fixed in 8fc006d: the expansion loop is now unbounded and ALWAYS queries at the configured cap before breaking (for(;;) that breaks only after a query, on candidates >= topK || k >= maxK) — for topK=3/maxK=64 the sequence is k = 6, 12, 24, 48, 64 (previously it stopped at 48, so rows ranked 49-64 were never examined). Locked by a deterministic regression: 3 target-pool rows + 60 other-pool rows, topK=3 scoped query returns exactly 3 in-scope rows for ANY distance order (bench/db-memory.test.ts).

File: engine/program.ts [hunk 13]
```diff
@@ -2815,53 +3020,108 @@ export const program = Effect.gen(function* () {
...
+              // never pays the Jupiter round-trip (when Jupiter is down that
+              // was a 10s timeout per ineligible pool per cycle); audit-reason
+              // precedence follows — pools failing local gates log the LOCAL
+              // reason, not an incidental token-risk one. Reuses the per-cycle
+              // token-risk cache, so a second consult costs no network call
+              // when the screening seam already fetched these mints.
+              if (config.jupiterTokenRiskEnabled !== false) {
+                const enterRisk = yield* Effect.promise(() =>
+                  consultTokenRisks([pool.tokenX, pool.tokenY], config),
+                );
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts]: Fixed in 9d81e8c: the token-risk consult is now the FINAL ENTER gate — it runs after the allocation gate, so allocation-dead pools (no headroom or MAX_OPEN_POSITIONS reached) never touch Jupiter (previously up to a 10s serial timeout per pool when requests failed). Audit precedence: allocation rejections log alloc-gate and consult nothing; token-risk rejections only fire for candidates that survived every local gate. Locked by a zero-fetch regression (decision-loop-screening.test.ts (14)).

File: engine/program.ts [hunk 14]
```diff
@@ -2013,14 +2013,133 @@ export const program = Effect.gen(function* () {
...
+          //   (a) Jupiter isSus → hard reject (aggregated RugCheck+Blockaid),
+          //   (b) Data API is_verified → exempt,
+          //   (c) Jupiter isVerified → pass via the overlay,
+          //   (d) otherwise the leg is unknown → smart-screening or strict reject.
+          // Trusted mints were already exempted above, so this only sees untrusted
+          // legs. Every branch logs its provenance; signals are never fabricated.
+          type LegStatus = "datapiVerified" | "sus" | "jupiterVerified" | "unknown";
+          const riskByMint: ReadonlyMap<string, TokenRiskSignal> =
+            config.jupiterTokenRiskEnabled !== false
+              ? yield* Effect.promise(() => consultTokenRisks([pool.tokenX, pool.tokenY], config))
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts:2088]: Fixed in adbdce89baa550f685e67e3dc7374ff71c31ba89: the local blacklist.checkPool now runs immediately after the mint-authority fetch and before the token-risk overlay consult — blacklisted pools reject deterministically with zero Jupiter calls (pinned by new tests 15/16, including smart-screening-on still rejecting blacklisted pools). is_blacklisted (Data API) order unchanged (already first); all fail-open semantics preserved.
